# BEE 4750 Homework 4: Linear Programming and Capacity Expansion

**Name**:

**ID**:

> **Due Date**
>
> Thursday, 11/07/23, 9:00pm

## Overview

### Instructions

-   Problem 1 asks you to formulate and solve a resource allocation
    problem using linear programming.
-   Problem 2 asks you to formulate, solve, and analyze a standard
    generating capacity expansion problem.
-   Problem 3 (5750 only) asks you to add a CO<sub>2</sub> constraint to
    the capacity expansion problem and identify changes in the resulting
    solution.

### Load Environment

The following code loads the environment and makes sure all needed
packages are installed. This should be at the start of most Julia
scripts.

In [1]:
import Pkg
Pkg.activate(@__DIR__)
Pkg.instantiate()

  Activating project at `c:\Users\aliap\Cornell\BEE 4750\hw4-trent-ali-1`


In [2]:
using JuMP
using HiGHS
using DataFrames
using Plots
using Measures
using CSV
using MarkdownTables

## Problems (Total: 50/60 Points)

### Problem 1 (20 points)

A farmer has access to a pesticide which can be used on corn, soybeans,
and wheat fields, and costs \$70/ha-yr to apply. The crop yields the
farmer can obtain following crop yields by applying varying rates of
pesticides to the field are shown in
<a href="#tbl-yields" class="quarto-xref">Table 1</a>.

| Application Rate (kg/ha) | Soybean (kg/ha) | Wheat (kg/ha) | Corn (kg/ha) |
|:------------------------:|:---------------:|:-------------:|:------------:|
|            0             |      2900       |     3500      |     5900     |
|            1             |      3800       |     4100      |     6700     |
|            2             |      4400       |     4200      |     7900     |

Table 1: Crop yields from applying varying pesticide rates for Problem
1.

The costs of production, *excluding pesticides*, for each crop, and
selling prices, are shown in
<a href="#tbl-costs" class="quarto-xref">Table 2</a>.

|   Crop   | Production Cost (\$/ha-yr) | Selling Price (\$/kg) |
|:--------:|:--------------------------:|:---------------------:|
| Soybeans |            350             |         0.36          |
|  Wheat   |            280             |         0.27          |
|   Corn   |            390             |         0.22          |

Table 2: Costs of crop production, excluding pesticides, and selling
prices for each crop.

Recently, environmental authorities have declared that farms cannot have
an *average* application rate on soybeans, wheat, and corn which exceeds
0.8, 0.7, and 0.6 kg/ha, respectively. The farmer has asked you for
advice on how they should plant crops and apply pesticides to maximize
profits over 130 total ha while remaining in regulatory compliance if
demand for each crop (which is the maximum the market would buy) this
year is 250,000 kg?

**In this problem**:

-   Formulate a linear program for this resource allocation problem,
    including clear definitions of decision variable(s) (including
    units), objective function(s), and constraint(s) (make sure to
    explain functions and constraints with any needed derivations and
    explanations). **Tip: Make sure that all of your constraints are
    linear**.
-   Implement the program in `JuMP.jl` and find the solution. How many
    ha should the farmer dedicate to each crop and with what pesticide
    application rate(s)? How much profit will the farmer expect to make?
-   The farmer has an opportunity to buy an extra 10 ha of land. How
    much extra profit would this land be worth to the farmer? Discuss
    why this value makes sense and whether you would recommend the
    farmer should make the purchase.

    


**ANSWER**

The below is formulated as a linear program as instructed. In the code, decision variables are labeled $Xj$ for j = 1-9 in order of appearence below. The program was formulated such that it could be run in JuMP.jl. Later, the program was re-solved such that 140 would be available and not 130. Answers to questions will be found in markdown sections and printed outputs.

---------------------------------------------------------------------------------------------------------------


DECISION VARIABLES: Area of Pesticide Application for each Crop, for each Application Rate (all in ha):

Rate = 0, Crop = Soybean: $x_{0,s}$ ; Rate = 0, Crop = Wheat: $x_{0,w}$ ; Rate = 0, Crop = Corn: $x_{0,c}$  

Rate = 1, Crop = Soybean: $x_{1,s}$ ; Rate = 1, Crop = Wheat: $x_{1,w}$ ; Rate = 1, Crop = Corn: $x_{1,c}$  

Rate = 2, Crop = Soybean: $x_{2,s}$ ; Rate = 2, Crop = Wheat: $x_{2,w}$ ; Rate = 2, Crop = Corn: $x_{2,c}$  

PROBLEM FORMULATION:

Crop Cost = 350 * ($x_{0,s}$ + $x_{1,s}$+$x_{2,s}$) + 280 * ($x_{0,w}$+$x_{1,w}$+$x_{2,w}$) + 390 * ($x_{0,c}$+$x_{1,c}$+$x_{2,c}$)

Pesticide Cost = 70 * (0*$x_{0,s}$ + 1*$x_{1,s}$ + 2*$x_{2,s}$ + 0*$x_{0,w}$ + 1*$x_{1,w}$ + 2*$x_{2,w}$ + 0*$x_{0,c}$ + 1*$x_{1,c}$ + 2*$x_{2,c}$)

Crop Revenue = 0.36 * (2900 $x_{0,s}$ + 3800 $x_{1,s}$ + 4400 $x_{2,s}$) + 0.27 * (3500 $x_{0,w}$ + 4100 $x_{1,w}$ + 4200 $x_{2,w}$) + 0.22 * (5900 $x_{0,c}$ + 6700 $x_{1,c}$ + 7900 $x_{2,c}$)

MAXIMIZE:

{$Revenue_{CROP}$ - $Cost_{CROP}$ - $Cost_{PEST}$}

SUBJECT TO:

$x_{0,s}$ , $x_{1,s}$ , $x_{2,s}$ , $x_{0,w}$ , $x_{1,w}$ , $x_{2,w}$ , $x_{0,c}$ , $x_{1,c}$ , $x_{2,c}$ $\geq$ 0

($x_{0,s}$ + $x_{1,s}$ + $x_{2,s}$ + $x_{0,w}$ + $x_{1,w}$ + $x_{2,w}$ + $x_{0,c}$ + $x_{1,c}$ + $x_{2,c}$) $\leq$ 130

($x_{1,s}$ + 2*$x_{2,s}$) - 0.8*($x_{0,s}$ + $x_{1,s}$ + $x_{2,s}$) $\leq$ 0

($x_{1,w}$ + 2*$x_{2,w}$) - 0.7*($x_{0,w}$ + $x_{1,w}$ + $x_{2,w}$) $\leq$ 0

($x_{1,c}$ + 2*$x_{2,c}$) - 0.6*($x_{0,c}$ + $x_{1,c}$ + $x_{2,c}$) $\leq$ 0

2900 $x_{0,s}$ + 3800 $x_{1,s}$ + 4400 $x_{2,s}$ $\leq$ 250000

3500 $x_{0,w}$ + 4100 $x_{1,w}$ + 4200 $x_{2,w}$ $\leq$ 250000

5900 $x_{0,c}$ + 6700 $x_{1,c}$ + 7900 $x_{2,c}$ $\leq$ 250000




In [3]:
farm_model = Model(HiGHS.Optimizer) # initialize model object
@variable(farm_model, X[1:9] >= 0) # non-negativity constraints

CropCost = 350 * (X[1] + X[2] + X[3]) + 280 * (X[4] + X[5] + X[6]) + 390 * (X[7] + X[8] + X[9])
PesticideCost = 70 * (X[2] + 2*X[3] + X[5] + 2*X[6] + X[8] + 2*X[9])
CropRevenue = 0.36 * (2900*X[1] + 3800*X[2] + 4400*X[3]) + 0.27 * (3500*X[4] + 4100*X[5] + 4200*X[6]) + 0.22 * (5900*X[7] + 6700*X[8] + 7900*X[9])

@objective(farm_model, Max, CropRevenue-CropCost-PesticideCost)

@constraint(farm_model, X[1]+X[2]+X[3]+X[4]+X[5]+X[6]+X[7]+X[8]+X[9] <= 130)
@constraint(farm_model, (X[2]+2*X[3])- 0.8*(X[1]+X[2]+X[3]) <= 0)
@constraint(farm_model, (X[5]+2*X[6])- 0.7*(X[4]+X[5]+X[6]) <= 0)
@constraint(farm_model, (X[8]+2*X[9])- 0.6*(X[7]+X[8]+X[9]) <= 0)
@constraint(farm_model, 2900*X[1] + 3800*X[2] + 4400*X[3] <= 250000)
@constraint(farm_model, 3500*X[4] + 4100*X[5] + 4200*X[6] <= 250000)
@constraint(farm_model, 5900*X[7] + 6700*X[8] + 7900*X[9] <= 250000)

print(farm_model)

In [4]:
optimize!(farm_model)

Running HiGHS 1.7.2 (git hash: 5ce7a2753): Copyright (c) 2024 HiGHS under MIT licence terms
Coefficient ranges:
  Matrix [2e-01, 8e+03]
  Cost   [7e+02, 1e+03]
  Bound  [0e+00, 0e+00]
  RHS    [1e+02, 2e+05]
Presolving model
7 rows, 9 cols, 27 nonzeros  0s
7 rows, 9 cols, 27 nonzeros  0s
Presolve : Reductions: rows 7(-0); columns 9(-0); elements 27(-0) - Not reduced
Problem not reduced by presolve: solving the LP
Using EKK dual simplex solver - serial
  Iteration        Objective     Infeasibilities num(sum)
          0    -1.8170297740e+02 Ph1: 7(28.3601); Du: 9(181.703) 0s
          7     1.1674116702e+05 Pr: 0(0) 0s
Model   status      : Optimal
Simplex   iterations: 7
Objective value     :  1.1674116702e+05
HiGHS run time      :          0.01


In [8]:
print("Soybean Tot: " * string(value.(X[1]) + value.(X[2]) + value.(X[3])) * " ha | ")
print("Wheat Tot: " * string(value.(X[4]) + value.(X[5]) + value.(X[6])) * " ha | ")
print("Corn Tot: " * string(value.(X[7]) + value.(X[8]) + value.(X[9])) * " ha")


Soybean Tot: 69.06077348066297 ha | Wheat Tot: 22.477688057798545 ha | Corn Tot: 38.46153846153848 ha

In [14]:
print("Soybean, P = 0, 1, 2 in ha: " * string(value.(X[1])) * ", " * string(value.(X[2])) * ", " * string(value.(X[3])))

Soybean, P = 0, 1, 2 in ha: 13.812154696132593, 55.24861878453038, 0.0

In [15]:
print("Wheat, P = 0, 1, 2 in ha: " * string(value.(X[4])) * ", " * string(value.(X[5])) * ", " * string(value.(X[6])))


Wheat, P = 0, 1, 2 in ha: 6.743306417339564, 15.73438164045898, 0.0

In [ ]:
print("Corn, P = 0, 1, 2 in ha: " * string(value.(X[7])) * ", " * string(value.(X[8])) * ", " * string(value.(X[9])))

Corn, P = 0, 1, 2: 26.923076923076927, 0.0, 11.538461538461554

In [17]:
print("Total Profit: " * string(objective_value(farm_model)))

Total Profit: 116741.16702082448

In [22]:
farm_model2 = Model(HiGHS.Optimizer) # initialize model object
@variable(farm_model2, X[1:9] >= 0) # non-negativity constraints

CropCost = 350 * (X[1] + X[2] + X[3]) + 280 * (X[4] + X[5] + X[6]) + 390 * (X[7] + X[8] + X[9])
PesticideCost = 70 * (X[2] + 2*X[3] + X[5] + 2*X[6] + X[8] + 2*X[9])
CropRevenue = 0.36 * (2900*X[1] + 3800*X[2] + 4400*X[3]) + 0.27 * (3500*X[4] + 4100*X[5] + 4200*X[6]) + 0.22 * (5900*X[7] + 6700*X[8] + 7900*X[9])

@objective(farm_model2, Max, CropRevenue-CropCost-PesticideCost)

@constraint(farm_model2, X[1]+X[2]+X[3]+X[4]+X[5]+X[6]+X[7]+X[8]+X[9] <= 140)
@constraint(farm_model2, (X[2]+2*X[3])- 0.8*(X[1]+X[2]+X[3]) <= 0)
@constraint(farm_model2, (X[5]+2*X[6])- 0.7*(X[4]+X[5]+X[6]) <= 0)
@constraint(farm_model2, (X[8]+2*X[9])- 0.6*(X[7]+X[8]+X[9]) <= 0)
@constraint(farm_model2, 2900*X[1] + 3800*X[2] + 4400*X[3] <= 250000)
@constraint(farm_model2, 3500*X[4] + 4100*X[5] + 4200*X[6] <= 250000)
@constraint(farm_model2, 5900*X[7] + 6700*X[8] + 7900*X[9] <= 250000)

optimize!(farm_model2)
print("Profit Gained with 10 more acres: " * string(objective_value(farm_model2)-objective_value(farm_model)))

Running HiGHS 1.7.2 (git hash: 5ce7a2753): Copyright (c) 2024 HiGHS under MIT licence terms
Coefficient ranges:
  Matrix [2e-01, 8e+03]
  Cost   [7e+02, 1e+03]
  Bound  [0e+00, 0e+00]
  RHS    [1e+02, 2e+05]
Presolving model
7 rows, 9 cols, 27 nonzeros  0s
7 rows, 9 cols, 27 nonzeros  0s
Presolve : Reductions: rows 7(-0); columns 9(-0); elements 27(-0) - Not reduced
Problem not reduced by presolve: solving the LP
Using EKK dual simplex solver - serial
  Iteration        Objective     Infeasibilities num(sum)
          0    -1.8170297740e+02 Ph1: 7(28.3601); Du: 9(181.703) 0s
          7     1.2403516702e+05 Pr: 0(0) 0s
Model   status      : Optimal
Simplex   iterations: 7
Objective value     :  1.2403516702e+05
HiGHS run time      :          0.00
Profit Gained with 10 more acres: 7293.999999999971

In [24]:
print("New Soybean Tot: " * string(value.(X[1]) + value.(X[2]) + value.(X[3])) * " ha | ")
print("New Wheat Tot: " * string(value.(X[4]) + value.(X[5]) + value.(X[6])) * " ha | ")
print("New Corn Tot: " * string(value.(X[7]) + value.(X[8]) + value.(X[9])) * " ha")

New Soybean Tot: 69.06077348066299 ha | New Wheat Tot: 32.47768805779855 ha | New Corn Tot: 38.461538461538446 ha

If the farmer were to buy 10 ha more of land, his profit would be $7,294 higher at $124,035.17 as shown in the outputs above. This makes sense because a more "relaxed" contraint on land allows for more land to be used for wheat in particular, which raises profit. I would recommend this choice if the price of the new land is less than $7,294 and there are no other negative externalities. However, with the little information given, it would be irresponsible to give a strong recommendation.


### Problem 2 (30 points)

For this problem, we will use hourly load (demand) data from 2013 in New
York’s Zone C (which includes Ithaca). The load data is loaded and
plotted below in <a href="#fig-demand" class="quarto-xref">Figure 1</a>.

In [ ]:
# load the data, pull Zone C, and reformat the DataFrame
NY_demand = DataFrame(CSV.File("data/2013_hourly_load_NY.csv"))
rename!(NY_demand, :"Time Stamp" => :Date)
demand = NY_demand[:, [:Date, :C]]
rename!(demand, :C => :Demand)
demand[:, :Hour] = 1:nrow(demand)

# plot demand
plot(demand.Hour, demand.Demand, xlabel="Hour of Year", ylabel="Demand (MWh)", label=:false)

Next, we load the generator data, shown in
<a href="#tbl-generators" class="quarto-xref">Table 3</a>. This data
includes fixed costs (\$/MW installed), variable costs (\$/MWh
generated), and CO<sub>2</sub> emissions intensity (tCO<sub>2</sub>/MWh
generated).

In [ ]:
gens = DataFrame(CSV.File("data/generators.csv"))

Finally, we load the hourly solar and wind capacity factors, which are
plotted in <a href="#fig-cf" class="quarto-xref">Figure 2</a>. These
tell us the fraction of installed capacity which is expected to be
available in a given hour for generation (typically based on the average
meteorology).

In [ ]:
# load capacify factors into a DataFrame
cap_factor = DataFrame(CSV.File("data/wind_solar_capacity_factors.csv"))

# plot January capacity factors
p1 = plot(cap_factor.Wind[1:(24*31)], label="Wind")
plot!(cap_factor.Solar[1:(24*31)], label="Solar")
xaxis!("Hour of the Month")
yaxis!("Capacity Factor")

p2 = plot(cap_factor.Wind[4344:4344+(24*31)], label="Wind")
plot!(cap_factor.Solar[4344:4344+(24*31)], label="Solar")
xaxis!("Hour of the Month")
yaxis!("Capacity Factor")

display(p1)
display(p2)

You have been asked to develop a generating capacity expansion plan for
the utility in Riley County, NY, which currently has no existing
electrical generation infrastructure. The utility can build any of the
following plant types: geothermal, coal, natural gas combined cycle gas
turbine (CCGT), natural gas combustion turbine (CT), solar, and wind.

While coal, CCGT, and CT plants can generate at their full installed
capacity, geothermal plants operate at maximum 85% capacity, and solar
and wind available capacities vary by the hour depend on the expected
meteorology. The utility will also penalize any non-served demand at a
rate of \$10,000/MWh.

**In this problem**:

-   Formulate a linear program for this capacity expansion problem,
    including clear definitions of decision variable(s) (including
    units), objective function(s), and constraint(s) (make sure to
    explain functions and constraints with any needed derivations and
    explanations).
-   Implement your linear program in `JuMP.jl`. Find the optimal
    solution. How much should the utility build of each type of
    generating plant? What will the total cost be? How much energy will
    be non-served?
-   What fraction of annual generation does each plant type produce? How
    does this compare to the breakdown of built capacity that you found
    in Problem 1.5? Do these results make sense given the generator
    data?
-   Make a plot of the electricity price in each hour. Discuss any
    trends that you see.

> **Significant Digits**
>
> Use `round(x; digits=n)` to report values to the appropriate
> precision! If your number is on a different order of magnitude and you
> want to round to a certain number of significant digits, you can use
> `round(x; sigdigits=n)`.

> **Getting Variable Output Values**
>
> `value.(x)` will report the values of a `JuMP` variable `x`, but it
> will return a special container which holds other information about
> `x` that is useful for `JuMP`. This means that you can’t use this
> output directly for further calculations. To just extract the values,
> use `value.(x).data`.

> **Suppressing Model Command Output**
>
> The output of specifying model components (variable or constraints)
> can be quite large for this problem because of the number of time
> periods. If you end a cell with an `@variable` or `@constraint`
> command, I *highly* recommend suppressing output by adding a
> semi-colon after the last command, or you might find that your
> notebook crashes.

### Problem 3 (10 points)

**This problem is only required for students in BEE 5750**.

The NY state legislature is considering enacting an annual
CO<sub>2</sub> limit, which for the utility would limit the emissions in
its footprint to 1.5 MtCO<sub>2</sub>/yr.

**In this problem**:

-   Reformulate your linear program from Problem 2 with any necessary
    changes to capture the CO<sub>2</sub> limit.
-   Implement the new optimization problem and find the optimal
    solution. How much should the utility build of each type of
    generating plant? What is different from your plan from Problem 1?
    Do these changes make sense?
-   What would the value to the utility be of allowing it to emit an
    additional 1000 tCO<sub>2</sub>/yr? An additional 5000?

## References

List any external references consulted, including classmates.